### PR TITLE
chore: Show FIDO devices in the chooser if allowed

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Refs CL:3309164

> This CL allows devices containing a top-level collection with a usage
from the FIDO usage page to appear in the WebHID device chooser if
the requesting origin is allowed to access FIDO.

Refs CL:3595839

> This CL makes sure users can see why a HID device does not show up in the browser picker to help debugging. See https://stackoverflow.com/a/71922823/422957 for context

Refs CL:3809913

> There is a smaller window between BrowserContext and HidService
destruction timing in service worker condition, which creates an UAF
issue as the stored BrowserContext pointer can still be accessed by the
HidService. This is fixed by using weak pointer of
ServiceWorkerContextCore to access BrowserContext when it is valid.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

**Review by jkleinsc — CHANGES_REQUESTED** [review]
This change removes functionality currently supported by Electron and replaces it with Chromium specific logic around allowing devices when a specific extension is installed.  I'm not sure I see the use case to do this.

**Comment by codebytere** [comment]
Changes addressed following offline conversation ☝️

**Comment by codebytere** [comment]
Changes addressed following offline conversation ☝️

File: shell/browser/hid/electron_hid_delegate.cc [hunk 1]
```diff
@@ -82,21 +85,28 @@ bool ElectronHidDelegate::HasDevicePermission(
     content::BrowserContext* browser_context,
     const url::Origin& origin,
     const device::mojom::HidDeviceInfo& device) {
-  return GetChooserContext(browser_context)
-      ->HasDevicePermission(origin, device);
+  auto* chooser_context = GetChooserContext(browser_context);
+  if (!chooser_context)
+    return false;
+  return chooser_context->HasDevicePermission(origin, device);
```
Comment (jkleinsc) [inline shell/browser/hid/electron_hid_delegate.cc:91]: We should more closely mirror upstream here, eg:
```suggestion
  return browser_context && GetChooserContext(browser_context->HasDevicePermission(origin, device);
```

File: shell/browser/hid/electron_hid_delegate.cc [hunk 2]
```diff
@@ -82,21 +85,28 @@ bool ElectronHidDelegate::HasDevicePermission(
...
 void ElectronHidDelegate::RevokeDevicePermission(
     content::BrowserContext* browser_context,
     const url::Origin& origin,
     const device::mojom::HidDeviceInfo& device) {
-  return GetChooserContext(browser_context)
-      ->RevokeDevicePermission(origin, device);
+  auto* chooser_context = GetChooserContext(browser_context);
+  if (!chooser_context)
+    return;
+  chooser_context->RevokeDevicePermission(origin, device);
```
Comment (jkleinsc) [inline shell/browser/hid/electron_hid_delegate.cc:101]: We should more closely mirror upstream here, eg:
```suggestion
  return browser_context && GetChooserContext(browser_context)->RevokeDevicePermission(origin, device);
```

File: shell/browser/hid/electron_hid_delegate.cc [hunk 3]
```diff
@@ -82,21 +85,28 @@ bool ElectronHidDelegate::HasDevicePermission(
...
 }
 
 device::mojom::HidManager* ElectronHidDelegate::GetHidManager(
     content::BrowserContext* browser_context) {
-  return GetChooserContext(browser_context)->GetHidManager();
+  auto* chooser_context = GetChooserContext(browser_context);
+  if (!chooser_context)
+    return nullptr;
+  return chooser_context->GetHidManager();
```
Comment (jkleinsc) [inline shell/browser/hid/electron_hid_delegate.cc:109]: We should more closely mirror upstream here, eg:
```suggestion
  if (!browser_context) {
    return nullptr;
  }
  return GetChooserContext(browser_context)->GetHidManager();
```

File: shell/browser/hid/electron_hid_delegate.cc [hunk 4]
```diff
@@ -117,14 +127,18 @@ const device::mojom::HidDeviceInfo* ElectronHidDelegate::GetDeviceInfo(
     content::BrowserContext* browser_context,
     const std::string& guid) {
   auto* chooser_context = GetChooserContext(browser_context);
+  if (!chooser_context)
+    return nullptr;
```
Comment (jkleinsc) [inline shell/browser/hid/electron_hid_delegate.cc:131]: This matches upstream, but if we keep it as is we need to change `GetChooserContext` to return nullptr if !browser_context.

File: shell/browser/hid/electron_hid_delegate.cc [hunk 5]
```diff
@@ -117,14 +127,18 @@ const device::mojom::HidDeviceInfo* ElectronHidDelegate::GetDeviceInfo(
...
 
 bool ElectronHidDelegate::IsFidoAllowedForOrigin(
     content::BrowserContext* browser_context,
     const url::Origin& origin) {
-  return base::CommandLine::ForCurrentProcess()->HasSwitch(
-      switches::kDisableHidBlocklist);
+  auto* chooser_context = GetChooserContext(browser_context);
+  if (!chooser_context)
+    return false;
+  return chooser_context->IsFidoAllowedForOrigin(origin);
```
Comment (jkleinsc) [inline shell/browser/hid/electron_hid_delegate.cc:141]: This kind of matches upstream, but if we keep it as is we need to change GetChooserContext to return nullptr if !browser_context.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Overall LGTM; just needs a few tweaks to more closely follow the changes in https://chromium-review.googlesource.com/c/chromium/src/+/3809913

**Review by jkleinsc — APPROVED** [review]

**Pull request merged**